### PR TITLE
add v1beta3 for Priority And Fairness

### LINKED ISSUE #104842: flowschemas resource is currently skipped by TestApplyResetFields test.
#### What happened:
`flowschemas` resource is currently skipped by `TestApplyResetFields` test.

https://github.com/kubernetes/kubernetes/blob/master/test/integration/apiserver/apply/reset_fields_test.go#L70-L74
```
var resetFieldsSkippedResources = map[string]struct{}{
	// TODO: flowschemas is flaking,
	// possible bug in the flowschemas controller.
	"flowschemas": {},
}
```

When it flakes, it fails with the following error
```
 reset_fields_test.go:288: expected conflict, got error <nil>
```

this is the test result:
```
--- FAIL: TestApplyResetFields (5.94s)
    --- FAIL: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1beta2,_Resource=flowschemas (0.05s)
    --- PASS: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1beta2,_Resource=prioritylevelconfigurations (0.01s)
    --- FAIL: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1beta1,_Resource=flowschemas (0.01s)
    --- PASS: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1beta1,_Resource=prioritylevelconfigurations (0.01s)
    --- PASS: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1alpha1,_Resource=flowschemas (0.01s)
    --- PASS: TestApplyResetFields/flowcontrol.apiserver.k8s.io/v1alpha1,_Resource=prioritylevelconfigurations (0.01s)
```

#### What you expected to happen:

We need to identify the root case of why it flakes and fix it. "flowschemas" being skipped means that we may not catch any errors related to field resetting when we add a new version of the API.

#### How to reproduce it (as minimally and precisely as possible):
- remove "flowschemas" from `resetFieldsSkippedResources`
- go test -v -count=1 ./test/integration/apiserver/apply -run="TestApplyResetFields/flowcontrol.apiserver.k8s.io"

run multiple times to see the flake

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### LINKED ISSUE #107574: Add patch annotations to APF API types in next version
As revealed in #104842 and #107547, the Condition slices in FlowSchemaStatus and PriorityLevelConfigurationStatus are missing the `patchStrategy` and `patchMergeKey` annotations that tell strategic merge patch how to handle the slice.

This should be fixed when we make the next version of this API.  See #107549 for exactly how.

@kubernetes/sig-api-machinery-misc 
@wojtek-t 
@deads2k 
@tkashem 
@lavalamp

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:
Fixes #107574
Fixes #104842

#### Does this PR introduce a user-facing change?
```release-note
Introduce v1beta3 for Priority and Fairness with the following changes to the API spec:
- rename 'assuredConcurrencyShares' (located under spec.limited') to 'nominalConcurrencyShares'
- apply strategic merge patch annotations to 'Conditions' of flowschemas and prioritylevelconfigurations
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controlplane/instance.go
Comment (tkashem) [inline pkg/controlplane/instance.go:640]: when do we remove `v1beta3`?

File: pkg/apis/flowcontrol/install/install.go
Comment (tkashem) [inline pkg/apis/flowcontrol/install/install.go:45]: Questions:
- if we remove `v1beta1`, that implies `v1alpha1` can also be removed, right?
- after removing `v1beta1` and `v1alpha1`, the serialization order will be `v1beta2, v1beta3` (1.26), and in 1.27 we will change it to `v1beta3, v1beta2`?

I am planning to address removal of `v1beta1` in a separate PR.

File: pkg/controlplane/instance.go
Comment (tkashem) [inline pkg/controlplane/instance.go:654]: @deads2k @liggitt  I think I need to add `flowcontrolv1beta3.SchemeGroupVersion` here as well after reading the KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/3136-beta-apis-off-by-default, 

But I had expected a test to fail due to this mis-configuration of the new beta version, I have added extra check to the test for this KEP:  please review https://github.com/kubernetes/kubernetes/pull/112402

Comment (deads2k) [reply pkg/controlplane/instance.go:654]: Yes, it should be added here and the resource enabled by name.

File: pkg/apis/flowcontrol/install/install.go
Comment (deads2k) [reply pkg/apis/flowcontrol/install/install.go:45]: we need to keep all serializations available because we do not require storage version migration.  We can now make the serialization version be v1beta2 because vN-1 understands that level.

File: test/integration/etcd/data.go
Comment (tkashem) [inline test/integration/etcd/data.go:317]: the tests are currently disabled
https://github.com/kubernetes/kubernetes/blob/d5e5f48b8805dcbd321675249bfba89fe997ff99/test/integration/apiserver/apply/reset_fields_test.go#L75-L79

we need https://github.com/kubernetes/kubernetes/pull/110173 to merge.

File: pkg/controlplane/storageversionhashdata/data.go
Comment (tkashem) [inline pkg/controlplane/storageversionhashdata/data.go:91]: changed, since serialization version is now `v1beta2`

File: pkg/apis/flowcontrol/install/install.go
Comment (cici37) [inline pkg/apis/flowcontrol/install/install.go:43]: Since there is a TODO added, would you mind to create an issue and get it tracked? Thank you

File: pkg/controlplane/instance.go
Comment (cici37) [inline pkg/controlplane/instance.go:641]: Could we add the life cycle comment here as well?

**Review by cici37 — COMMENTED** [review]
I guess updating the expectedGVK in test/integration/etcd/data.go should fix the failing test. Thank you!

**Comment by tkashem** [comment]
> I guess updating the expectedGVK in test/integration/etcd/data.go should fix the failing test. Thank you!

`TestEtcdStoragePath` is failing with the following error
```
=== RUN   TestEtcdStoragePath/flowcontrol.apiserver.k8s.io/v1beta3,_Resource=flowschemas
    etcd_storage_path_test.go:193: Previous releases understand non-alpha versions ["v1beta1"], but do not understand the expected current storage version "v1beta2". This means a current server will store data in etcd that is not understood by a previous version.
```

I checked:
- https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/api/testdata/v1.24.0
-  https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/api/testdata/v1.25.0

I don't see serialized `v1beta2` for `flowcontrol.apiserver.k8s.io` in testdata/v1.24.0 or testdata/v1.25.0.  `v1beta2` was added in 1.23, did we miss something there? 

Is the solution to add `flowcontrol.apiserver.k8s.io/v1beta2` in `testdata/v1.25.0`?

**Comment by liggitt** [comment]
> I don't see serialized `v1beta2` for `flowcontrol.apiserver.k8s.io` in testdata/v1.24.0 or testdata/v1.25.0. `v1beta2` was added in 1.23, did we miss something there?

https://github.com/kubernetes/kubernetes/blob/release-1.25/staging/src/k8s.io/api/roundtrip_test.go#L52 is missing v1beta2 ... I regenerated those a v1.24.0 and v1.25.0 and HEAD and opened https://github.com/kubernetes/kubernetes/pull/112487 to add them.

File: pkg/apis/flowcontrol/install/install.go
Comment (tkashem) [reply pkg/apis/flowcontrol/install/install.go:43]: sure, please see https://github.com/kubernetes/kubernetes/issues/112512

File: pkg/controlplane/instance.go
Comment (tkashem) [reply pkg/controlplane/instance.go]: I added the comment, but can we actually remove `v1beta*` since these are legacy?

**Comment by tkashem** [comment]
> I guess v1alpha1 and v1beta1 can both be removed in 1.26 and it is better to be handled in a separate PR.

@cici37 can we actually remove v1beta1 or v1alpha1? There might be existing cluster(s) where existing apf object(s) are stored in v1alpha1/v1beta1.

File: pkg/controlplane/instance.go
Comment (liggitt) [inline pkg/controlplane/instance.go:642]: v1beta3 is to be removed in 1.32 (deprecated in 1.29, removed in 1.32)

File: pkg/apis/flowcontrol/install/install.go
Comment (liggitt) [reply pkg/apis/flowcontrol/install/install.go:43]: thanks, can you put that issue number in the TODO comment to close the loop?

File: cmd/kube-apiserver/app/options/validation.go
Comment (liggitt) [inline cmd/kube-apiserver/app/options/validation.go:130]: since the internal controller uses v1beta3, I think we need that specific version enabled, and should drop v1beta2 and v1beta1 from this list

File: pkg/apis/flowcontrol/v1beta3/defaults.go
Comment (liggitt) [inline pkg/apis/flowcontrol/v1beta3/defaults.go:35]: the v1beta3 boundary is a chance to modify defaults from v1beta2 if needed... are there any changes to defaults that have been contemplated in the past that should be done at this point?

**Comment by liggitt** [comment]
thanks for the review @cici37 

API mechanics lgtm as well, just a couple cleanup nits and one question about whether any changes to defaults are desired as part of this promotion

**Comment by liggitt** [comment]
I just noticed https://github.com/kubernetes/kubernetes/issues/107574 which looks like an API change queued up for the next APF version. Should that be done in this PR? Are there other changes like that queued up?

**Comment by liggitt** [comment]
Make the patch annotation change in this PR, so it's super clear there are no clients built with the previous annotations.

The field addition can happen in a follow up PR

**Comment by tkashem** [comment]
```


1/6570 Tests Failed. | expand_less
-- | --
Kubernetes e2e suite: [It] [sig-node] Pods Extended Pod Container lifecycle should not create extra sandbox if all containers are done expand_less1m13s{Sep 16 13:32:45.104: timed out waiting for the condition failed test/e2e/node/pods.go:276 k8s.io/kubernetes/test/e2e/node.glob..func10.4.2() 	test/e2e/node/pods.go:276 +0x5dc} | Kubernetes e2e suite: [It] [sig-node] Pods Extended Pod Container lifecycle should not create extra sandbox if all containers are done expand_less | 1m13s | {Sep 16 13:32:45.104: timed out waiting for the condition failed test/e2e/node/pods.go:276 k8s.io/kubernetes/test/e2e/node.glob..func10.4.2() 	test/e2e/node/pods.go:276 +0x5dc}
Kubernetes e2e suite: [It] [sig-node] Pods Extended Pod Container lifecycle should not create extra sandbox if all containers are done expand_less | 1m13s
{Sep 16 13:32:45.104: timed out waiting for the condition failed test/e2e/node/pods.go:276 k8s.io/kubernetes/test/e2e/node.glob..func10.4.2() 	test/e2e/node/pods.go:276 +0x5dc}
```

/test pull-kubernetes-e2e-kind-ipv6

File: cmd/kube-apiserver/app/options/validation.go
Comment (tkashem) [reply cmd/kube-apiserver/app/options/validation.go:130]: for 1.26, we are setting `v1beta2` as the storage version, we should not allow the user to disable `v1beta2` when APF is enabled, or is it orthogonal to the storage version?

Comment (liggitt) [reply cmd/kube-apiserver/app/options/validation.go:130]: this controls what gets served, not stored

File: pkg/apis/flowcontrol/v1beta3/defaults.go
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1beta3/defaults.go:35]: We do not have any queued up.

File: pkg/controlplane/instance.go
Comment (MikeSpreitzer) [reply pkg/controlplane/instance.go:642]: +1

**Review by MikeSpreitzer — COMMENTED** [review]
Looks good to me except for that comment oversight.

File: test/integration/etcd/data.go
Comment (tkashem) [reply test/integration/etcd/data.go]: yes, I set the ExpectedGVK to `v1beta2`

File: pkg/apis/flowcontrol/install/install.go
Comment (tkashem) [reply pkg/apis/flowcontrol/install/install.go]: Yes, done.

File: pkg/controlplane/instance.go
Comment (tkashem) [reply pkg/controlplane/instance.go]: done

File: pkg/controlplane/instance.go
Comment (tkashem) [reply pkg/controlplane/instance.go]: thanks, comments updated

File: cmd/kube-apiserver/app/options/validation.go
Comment (tkashem) [reply cmd/kube-apiserver/app/options/validation.go]: done, I have added a unit test for that as well.

**Comment by liggitt** [comment]
changes lgtm, if the field rename mentioned in https://github.com/kubernetes/kubernetes/issues/107574#issuecomment-1249564952 is planned for v1beta3, that should happen in this PR as well

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm

**Comment by MikeSpreitzer** [comment]
/remove-lgtm
for https://github.com/kubernetes/kubernetes/pull/112306#issuecomment-1251357849

**Comment by tkashem** [comment]
> &errors.errorString{s:"http: Handler timeout"}: http: Handler timeout

> I0921 23:16:33.306584   53214 trace.go:205] Trace[111769827]: "List(recursive=true) etcd3" audit-id:5b9bdbe4-e87f-4114-a074-0ebb7d5ce8d9,key:/leases/namespace-1663802066-11177,resourceVersion:,resourceVersionMatch:,limit:0,continue: (21-Sep-2022 23:16:21.484) (total time: 11822ms):
Trace[111769827]: [11.822201693s] [11.822201693s] END

handler timeout and etcd slowness, does not seem to be related to this PR

/test pull-kubernetes-integration

File: pkg/apis/flowcontrol/v1alpha1/conversion.go
Comment (liggitt) [inline pkg/apis/flowcontrol/v1alpha1/conversion.go:27]: for these manual conversion methods, follow this pattern:

```
func Convert_... {
  if err := autoConvert...(in, out, s); err != nil {
    return err
  }
  ... manual field conversion ... 
  return nil
}
```

Comment (liggitt) [reply pkg/apis/flowcontrol/v1alpha1/conversion.go:27]: applies to all the manual methods

File: pkg/apis/flowcontrol/validation/validation.go
Comment (liggitt) [inline pkg/apis/flowcontrol/validation/validation.go:389]: because the field name differs by version, you'll need to pass in the request version and use the appropriate field name based on the request version.

See https://github.com/kubernetes/kubernetes/pull/111506/files as an example

Comment (tkashem) [reply pkg/apis/flowcontrol/validation/validation.go]: Thanks for the pointer, I was going to ask about that.

File: pkg/apis/flowcontrol/validation/validation.go
Comment (tkashem) [inline pkg/apis/flowcontrol/validation/validation.go:357]: @MikeSpreitzer this function is not used anywhere, so I removed it.

**Comment by tkashem** [comment]
> {Failed  === RUN   TestCreateHealthcheck/timeouts_if_response_time_higher_than_default_timeout
>    factory_test.go:141: healthcheck() missmatch want context deadline exceeded got etcd client connection not yet established
>    --- FAIL: TestCreateHealthcheck/timeouts_if_response_time_higher_than_default_timeout (1.00s)
> }

/test pull-kubernetes-unit

**Comment by tkashem** [comment]
it's ready for another pass, thanks!

File: pkg/apis/flowcontrol/validation/validation.go
Comment (tkashem) [reply pkg/apis/flowcontrol/validation/validation.go:394]: does `getVersionedFieldNameForConcurrencyShares` read ok?

File: pkg/apis/flowcontrol/validation/validation.go
Comment (tkashem) [reply pkg/apis/flowcontrol/validation/validation.go]: Just curious as to why we do typed validation on the internal object, could we not validate on the typed request object? (I am not proposing any change, just want to understand the reason or historical context behind it)

Comment (liggitt) [reply pkg/apis/flowcontrol/validation/validation.go]: so that validation is consistent no matter what version the request came in via (or where there are differences, they are intentional and clearly visible). As complex as some of our validation is, requiring replicating it against multiple external versions would surely drift and introduce validation gaps via some versions

Comment (tkashem) [reply pkg/apis/flowcontrol/validation/validation.go]: so we traded off additional complexity in the validation logic to avoid the drift, thanks for the explanation!

File: pkg/apis/flowcontrol/validation/validation.go
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/validation/validation.go:394]: yes

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM
Thanks!

**Review by cici37 — COMMENTED** [review]
Couple questions:
1. Since the patch annotation is added and the controller switched to use v1beta3, should [[reset_fields_test.go](https://github.com/kubernetes/kubernetes/pull/112306/commits/66fc0d703794f309c9715028d3b63f64c281a5fd#diff-8c748d5484a717f07318d05bccbd0deb4989c74eecb466aed9620e35581d9754)](https://github.com/kubernetes/kubernetes/blob/c012d901d8bee86ef3e3c9472a1a4a0368a34775/test/integration/apiserver/apply/reset_fields_test.go#L73-L77) updated now?
2. Maybe worth to keep in mind(Could be added in a followup as well): [test/integration/apiserver/flowcontrol/fs_condition_test.go](https://github.com/MikeSpreitzer/kubernetes/blob/ae3c3b0fd06efdfb464384aeb395cf777dc451b9/test/integration/apiserver/flowcontrol/fs_condition_test.go)
3. If we plan to address 1, maybe mention this PR will fix https://github.com/kubernetes/kubernetes/issues/104842
4. Could be a followup PR: update lifecycle directives on old flowcontrol versions to point to new v1beta3 version
e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/flowcontrol/v1beta1/types.go#L108 and v1alpha1 and v1beta2 types

Thank you! :)

**Comment by liggitt** [comment]
thanks for the API review, marking API changes as approved

looks like dropping the exclusion from reset_fields_test.go and making sure the test doesn't flake is the only outstanding thing for this PR

**Comment by tkashem** [comment]
> looks like dropping the exclusion from reset_fields_test.go and making sure the test doesn't flake is the only outstanding thing for this PR

I have an open PR https://github.com/kubernetes/kubernetes/pull/112575 for that. I also tested locally.

stress test on `TestApplyResetFields`, on master (`flowschemas` is skipped) - 1000 runs and 0 failure
```
$ go test -c ./test/integration/apiserver/apply/
$ stress ./apply.test -test.run=TestApplyResetFields
5s: 0 runs so far, 0 failures
10s: 5 runs so far, 0 failures
...
18m50s: 1014 runs so far, 0 failures
```

stress test on `TestApplyResetFields`, on master (`flowschemas` is not skipped): 2 failures within first 8 runs
```
$ stress ./apply.test -test.run=TestApplyResetFields
5s: 0 runs so far, 0 failures
10s: 4 runs so far, 1 failures (25.00%)
15s: 8 runs so far, 2 failures (25.00%)
```

stress test on `TestApplyResetFields`, on this PR (`flowschemas` is not skipped): 1000 runs, no failure
```
$ stress ./apply.test -test.run=TestApplyResetFields
5s: 0 runs so far, 0 failures
10s: 2 runs so far, 0 failures
...
21m35s: 1101 runs so far, 0 failures
21m40s: 1104 runs so far, 0 failures
```

Once we merge #112575, I will keep an eye on CI just to make sure we don't see the flakes.

**Comment by liggitt** [comment]
sounds good, can fast-follow this PR with https://github.com/kubernetes/kubernetes/pull/112575

**Comment by tkashem** [comment]
@cici37 thanks for the review!

> Since the patch annotation is added and the controller switched to use v1beta3, ) should reset_fields_test.go be updated now?

Yes, https://github.com/kubernetes/kubernetes/pull/112575 

> Could be a followup PR: update lifecycle directives on old flowcontrol versions to point to new v1beta3 version

https://github.com/kubernetes/kubernetes/pull/112832

**Comment by wojtek-t** [comment]
Great to see it merged - thanks!

**Pull request merged**
